src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -140,6 +141,43 @@ exception StandardOperatorRedefinitionWarning of string * range
 
 exception InvalidInternalsVisibleToAssemblyName of badName: string * fileName: string option
 
+//----------------------------------------------------------------------------------------------
+// Helpers for determining if/what specifiers a string has.
+// Used to decide if interpolated string can be lowered to a concat call.
+// We don't care about single- vs multi-$ strings here, because lexer took care of that already.
+//----------------------------------------------------------------------------------------------
+[<return: Struct>]
+let (|HasFormatSpecifier|_|) (s: string) =
+    if
+        Regex.IsMatch(
+            s,
+            // Regex pattern for something like: %[flags][width][.precision][type]
+            """
+            (^|[^%])                # Start with beginning of string or any char other than '%'
+            (%%)*%                  # followed by an odd number of '%' chars
+            [+-0 ]{0,3}             # optionally followed by flags
+            (\d+)?                  # optionally followed by width
+            (\.\d+)?                # optionally followed by .precision
+            [bscdiuxXoBeEfFgGMOAat] # and then a char that determines specifier's type
+            """,
+            RegexOptions.Compiled ||| RegexOptions.IgnorePatternWhitespace)
+    then
+        ValueSome HasFormatSpecifier
+    else
+        ValueNone
+
+// Removes trailing "%s" unless it was escaped by another '%' (checks for odd sequence of '%' before final "%s")
+let (|WithTrailingStringSpecifierRemoved|) (s: string) =
+    if s.EndsWith "%s" then
+        let i = s.AsSpan(0, s.Length - 2).LastIndexOfAnyExcept '%'
+        let diff = s.Length - 2 - i
+        if diff &&& 1 <> 0 then
+            s[..i]
+        else
+            s
+    else
+        s
```

Comment:
Not necessarily for this particular PR, but I wonder if we can/should accept/return spans here, to not copy strings.

Reply:
I might be confusing something here, but this only creates a copy of a string if we are slicing it (in `s[..i]` on line 175), right?
If this function accepted and returned a span, we would still have to convert back to a string when calling `mkString`, so would we gain anything?
Am I wrong here, or are you suggesting to change the other APIs to accept spans as well?

Reply:
I would suggest changing majority of our non-public APIs to work over spans eventually.

Reply:
> There are literally 0 reasons of using strings over of spans where we can.

Spans are great, but the ergonomics aren't always as good right now — we don't yet support pattern-matching on spans, for example, and you can't return a span from a partial active pattern (you can't do `Some span`/`ValueSome span`).

Now, if we ever get around to type-directed `[]` and a spread operator...

```fsharp
match "abc123".AsSpan 1 with
| ['b'; 'c'; '1'; ..] -> …
| _ -> …
```

or even just being able to match spans against string or array literals (by, say, invoking `op_Implicit`)

```fsharp
match "abc123".AsSpan (1, 3) with
| "bc1" -> …
| _ -> …
```

— _then_ it's really over for strings :)

Reply:
Matching on strings or sequences in compiler is mostly on constants, and not something we return as substring of something usually. I'm mostly talking about cases in checker when on checks we need to return a slice of something, and no we just allocate string. We don't need to in most of the cases, I think.

Reply:
> We don't need to in most of the cases, I think.

Yes, in general I agree that there are a lot of places where spans could be used instead of strings.

Reply:
(I went ahead and logged https://github.com/fsharp/fslang-suggestions/issues/1351 anyway 🙃)

Reply:
I think I will opt not to change it to span for now - in addition to pattern matching, it seems there are some problems using spans with regular expressions (in netstandard2.0). I'd rather tackle this when we decide to change our syntax and typed tree expression to use spans (I really really doubt that usage of these two functions will grow to the point where it is a problem)